src/services/ServiceLocator.hh

```diff
@@ -0,0 +1,129 @@
+#ifndef RA_SERVICE_LOCATOR_HH
+#define RA_SERVICE_LOCATOR_HH
+#pragma once
+
+#include <assert.h>
+#include <memory>
+
+namespace ra {
+namespace services {
+
+class ServiceLocator
+{
+public:
+    /* returns a referece to a service implementing the requested interface
+     * 
+     * @throw std::runtime_error if no service is provided for the requested interface
+     */
+    template <class TClass>
+    static const TClass& Get()
+    {
+        return Service<TClass>::Get();
+    }
```

Comment:
It's a static method that returns a const reference, why would I use constexpr? (which doesn't compile by the way):
```
2>e:\source\raintegration\src\ra_leaderboard.cpp(156): error C2352: 'ra::services::ServiceLocator::Get': illegal call of non-static member function
2>e:\source\raintegration\src\services\servicelocator.hh(19): note: see declaration of 'ra::services::ServiceLocator::Get'
```
const data is supposed to be immutable, and therefore can safely be shared across threads.

`GetMutable<>` returns a non-const version that may not be valid across threads, but the threading logic should be internal.

If the concern is about the pointer itself not being thread-safe - it's not expected that the pointers will change once they're initialized. That would cause all sorts of problems if classes captured references to the services (a la dependency injection).

Reply:
Not a big deal, it doesn't make sense to put constexpr in interfaces anyway.